File: go/mandelbrot.go
package main

import "math"
import "os"
import "fmt"
import "time"

const parallelism = 16

// ********************************************************************
// Mandelbrot
const N = 1000        // grid size
const depth = 200     // max iterations
const escape2 = 400.0 // escape radius ^ 2

func trans_x(x int) float64 {
	return 3.0*(float64(x)/N) - 2.0
}

func trans_y(y int) float64 {
	return 3.0*(float64(y)/N) - 1.5
}

func mag2(x complex128) float64 {
	return float64(real(x)*real(x) + imag(x)*imag(x))
}

func mandel(idx int) float64 {
	var z0 complex128 = complex(trans_x(idx%N), trans_y(idx/N))
	var z complex128 = complex(0.0, 0.0)
	var magz2 float64
	var k uint

	for k = 0; k < depth; k++ {
		z = z*z + z0
		magz2 = mag2(z)
		if magz2 >= escape2 {
			break
		}
	}
	return math.Log(float64(k) + 1.0 - math.Log(math.Log(math.Max(magz2, escape2))/2.0)/math.Log(2.0))
}

// ********************************************************************
// Color mapping
// @todo Make const
var color_map = [][3]float64{
	{0.0, 0.0, 0.5},
	{0.0, 0.0, 1.0},
	{0.0, 0.5, 1.0},
	{0.0, 1.0, 1.0},
	{0.5, 1.0, 0.5},
	{1.0, 1.0, 0.0},
	{1.0, 0.5, 0.0},
	{1.0, 0.0, 0.0},
	{0.5, 0.0, 0.0},
	{0.5, 0.0, 0.0},
	{1.0, 0.0, 0.0},
	{1.0, 0.5, 0.0},
	{1.0, 1.0, 0.0},
	{0.5, 1.0, 0.5},
	{0.0, 1.0, 1.0},
	{0.0, 0.5, 1.0},
	{0.0, 0.0, 1.0},
	{0.0, 0.0, 0.5},
	{0.0, 0.0, 0.0}}

func interpolate(d float64, v0 float64, v1 float64) uint {
	return uint((d*(v1-v0) + v0) * 255.0)
}

var min_result float64 = 0.0
var max_result float64 = 0.0

const stops = 18

func map_to_argb(x float64) uint {
	x = (x - min_result) / (max_result - min_result) * stops
	var bin = uint32(x)
	if bin >= stops {
		return 0xff000000
	} else {
		r0 := color_map[bin][0]
		g0 := color_map[bin][1]
		b0 := color_map[bin][2]
		r1 := color_map[bin+1][0]
		g1 := color_map[bin+1][1]
		b1 := color_map[bin+1][2]
		d := x - float64(bin)
		r := interpolate(d, r0, r1)
		g := interpolate(d, g0, g1)
		b := interpolate(d, b0, b1)
		return b | (g << 8) | (r << 16) | 0xff000000
	}
}

var log_count [N * N]float64
var argb_array [N * N]uint

func do_mandel_range(beg int, end int) {
	for idx := beg; idx < end; idx++ {
		log_count[idx] = mandel(idx)
	}
}

func do_map_to_argb_range(beg int, end int) {

	for i := beg; i < end; i++ {
		argb_array[i] = map_to_argb(log_count[i])
	}
}

const job_size = N * N / parallelism
const worker_threads = parallelism

func do_mandel() {
	if parallelism > 1 {
		results := make([]chan int, worker_threads)

		for i := 0; i < worker_threads; i++ {
			results[i] = make(chan int)
			go func(n int) {
				do_mandel_range(job_size*n, job_size*(n+1))
				results[n] <- 1
			}(i)
		}

		for i := 0; i < worker_threads; i++ {
			<-results[i]
		}
	} else {
		do_mandel_range(0, N*N)
	}
}

func do_map_to_argb() {
	min_result = math.MaxFloat64
	max_result = math.SmallestNonzeroFloat64
	for _, val := range log_count {
		min_result = math.Min(min_result, val)
		max_result = math.Max(max_result, val)
	}

	if parallelism > 1 {
		results := make([]chan int, worker_threads)

		for i := 0; i < worker_threads; i++ {
			results[i] = make(chan int)
			go func(n int) {
				do_map_to_argb_range(job_size*n, job_size*(n+1))
				results[n] <- 1
			}(i)
		}

		for i := 0; i < worker_threads; i++ {
			<-results[i]
		}
	} else {
		do_map_to_argb_range(0, N*N)
	}
}

func main() {

	time_start := time.Now()

	do_mandel()
	do_map_to_argb()

	time_stop := time.Now()

	time_elapsed := time_stop.Sub(time_start)
	fmt.Printf("elapsed: %v\n", time_elapsed)

	// Save image file

	imgfile, _ := os.Create("out.ppm")
	imgfile.WriteString(fmt.Sprintf("P6\n%v %v\n255\n", N, N))

	// Extract RGB from RGBA
	for i := 0; i < N*N; i++ {
		rgba := argb_array[i]
		r := uint8((rgba & 0x00ff0000) >> 16)
		g := uint8((rgba & 0x0000ff00) >> 8)
		b := uint8((rgba & 0x000000ff))
		rgb_out := make([]byte, 3)
		rgb_out[0] = r
		rgb_out[1] = g
		rgb_out[2] = b
		imgfile.Write(rgb_out)
	}
}

File: rust/mandelbrot.rs

#![allow(unused_must_use)]

extern crate num;
extern crate time;

use std::num::ln;
use std::num::Bounded;
use num::complex::{Cmplx,Complex64};
use time::precise_time_ns;
use std::io::File;
use std::path::Path;

//static parallelism: uint = 16;
static parallelism: uint = 1;

// ********************************************************************
// Mandelbrot
static N: uint = 1000;      // grid size
static depth: uint = 200;   // max iterations
static escape2: f64 = 400.0; // escape radius ^ 2

#[inline]
fn trans_x(x: uint) -> f64 {
    3.0 * (x as f64 / N as f64) - 2.0
}

#[inline]
fn trans_y(y: uint) -> f64 {
    3.0 * (y as f64 / N as f64) - 1.5
}

#[inline]
fn mandel(idx: uint) -> f64 {
    let z0: Complex64 = Cmplx::new(trans_x(idx % N), trans_y(idx / N));
    let mut z: Complex64 = Cmplx::new(0.0, 0.0);
    let mut k: uint = 0;
    let mut magz2: f64 = 0.0;

    for kn in range(0, depth) {
        k = kn;
        z = z * z + z0;

        magz2 = z.norm_sqr();
        if magz2 >= escape2 {
            break;
        }
    }

    let v = if magz2 > escape2 { magz2 } else { escape2 };

    ln((k+1) as f64 + 1.0 - ln(ln(v) / 2.0) / ln(2.0))
}

// ********************************************************************
// Color mapping
static color_map: [[f32, ..3], ..19] =
    [ [ 0.0, 0.0, 0.5 ] ,
      [ 0.0, 0.0, 1.0 ] ,
      [ 0.0, 0.5, 1.0 ] ,
      [ 0.0, 1.0, 1.0 ] ,
      [ 0.5, 1.0, 0.5 ] ,
      [ 1.0, 1.0, 0.0 ] ,
      [ 1.0, 0.5, 0.0 ] ,
      [ 1.0, 0.0, 0.0 ] ,
      [ 0.5, 0.0, 0.0 ] ,
      [ 0.5, 0.0, 0.0 ] ,
      [ 1.0, 0.0, 0.0 ] ,
      [ 1.0, 0.5, 0.0 ] ,
      [ 1.0, 1.0, 0.0 ] ,
      [ 0.5, 1.0, 0.5 ] ,
      [ 0.0, 1.0, 1.0 ] ,
      [ 0.0, 0.5, 1.0 ] ,
      [ 0.0, 0.0, 1.0 ] ,
      [ 0.0, 0.0, 0.5 ] ,
      [ 0.0, 0.0, 0.0 ] ];

#[inline]
fn interpolate(d: &f32, v0: &f32, v1: &f32) -> uint {
    ((*d * (*v1 - *v0) + *v0) * 255.0).round() as uint
}

static mut min_result: f64 = 0.0;
static mut max_result: f64 = 0.0;

static stops: uint = 18;

#[inline]
fn map_to_argb(x: f64) -> uint {
    let mut x1: f64 = 0.0;
    unsafe {x1 = (x - min_result) / (max_result - min_result) * (stops as f64);}
    let bin: uint = x1 as uint;
    if bin >= stops {
        0xff000000
    } else {
        let r0: &f32 = &color_map[bin][0];
        let g0: &f32 = &color_map[bin][1];
        let b0: &f32 = &color_map[bin][2];
        let r1: &f32 = &color_map[bin+1][0];
        let g1: &f32 = &color_map[bin+1][1];
        let b1: &f32 = &color_map[bin+1][2];
        let d = x1 as f32 - bin as f32;
        let r = interpolate(&d, r0, r1);
        let g = interpolate(&d, g0, g1);
        let b = interpolate(&d, b0, b1);
        b | (g << 8) | (r << 16) | 0xff000000
    }
}

static mut log_count: [f64, ..N*N] = [0.0, .. N*N];
static mut argb_array: [uint, ..N*N] = [0, .. N*N];

fn do_mandel_range(beg: uint, end: uint) {
    for idx in range(beg, end) {
        unsafe { log_count[idx] = mandel(idx);}
    }
}

fn do_map_to_argb_range(beg: uint, end: uint) {
    for i in range(beg, end) {
        unsafe {
            argb_array[i] = map_to_argb(log_count[i]);
        }
    }
}

static job_size: uint = N*N / parallelism;
static worker_threads: uint = parallelism - 1;

fn do_mandel() {
    if parallelism > 1 {
/*
        QFuture<void> results[worker_threads];
        for(int i = 0; i < worker_threads; ++i)
            results[i] = QtConcurrent::run(do_mandel_range, job_size * i, job_size * (i + 1));

        do_mandel_range(job_size * worker_threads, N*N);

        for(int i = 0; i < worker_threads; ++i)
            results[i].waitForFinished();
*/
    }
    else {
        do_mandel_range(0, N*N);
    }
}

fn do_map_to_argb() {

    // We can't use these because TotalOrd is not defined for f32/f64
    //    min_result = log_count.iter().min();
    //    max_result = log_count.iter().max();

    unsafe { // mut static
        min_result = Bounded::max_value();
        max_result = Bounded::min_value();
        for val in log_count.iter() {
            if *val < min_result {
                min_result = *val;
            }
            if *val > max_result {
                max_result = *val;
            }
        }
    }

    if parallelism > 1
    {
/*
        QFuture<void> results[worker_threads];
        for(int i = 0; i < worker_threads; ++i)
            results[i] = QtConcurrent::run(do_map_to_argb_range, job_size * i, job_size * (i + 1));

        do_map_to_argb_range(job_size * worker_threads, N*N);

        for(int i = 0; i < worker_threads; ++i)
            results[i].waitForFinished();
*/
    } else {
        do_map_to_argb_range(0, N*N);
    }
}

fn main() {

    let time_start = precise_time_ns();

    do_mandel();
    do_map_to_argb();

    let time_stop = precise_time_ns();

    let time_elapsed_ms = (time_stop - time_start) as f32 / 1000000.0;

    println!("Elapsed: {} ms", time_elapsed_ms);

    // Save image

    // Header
    let mut file = File::create(&Path::new("out.ppm"));
    file.write(bytes!("P6\n"));
    file.write_str(format!("{} {}\n255\n", N, N));

    // Extract RGB from RGBA
    for i in range(0, N*N) {
        let mut rgba: u32;
        unsafe {
            rgba = argb_array[i] as u32;
        }
        let r: u8 = ((rgba & 0x00ff0000) >> 16) as u8;
        let g: u8 = ((rgba & 0x0000ff00) >>  8) as u8;
        let b: u8 = ((rgba & 0x000000ff))       as u8;
        file.write_u8(r);
        file.write_u8(g);
        file.write_u8(b);
    }
}
